nav2_costmap_2d/plugins/semantic_segmentation_layer.cpp

```diff
@@ -0,0 +1,304 @@
+  // get the latest segmentations thread safely
+  segmentation_buffer_->lock();
+  segmentation_buffer_->getSegmentations(segmentations);
+  segmentation_buffer_->unlock();
+  for (auto& segmentation : segmentations)
+  {
+    if (debug_topics_)
+    {
+      proc_pointcloud_pub_->publish(*segmentation.cloud_);
+    }
+    sensor_msgs::PointCloud2ConstIterator<float> iter_x(*segmentation.cloud_, "x");
+    sensor_msgs::PointCloud2ConstIterator<float> iter_y(*segmentation.cloud_, "y");
+    sensor_msgs::PointCloud2ConstIterator<uint8_t> iter_class(*segmentation.cloud_, "class");
+    // For now, confidence of the inference is not used to adjust cost
+    // sensor_msgs::PointCloud2ConstIterator<uint8_t> iter_confidence(*segmentation.cloud_,
+    // "confidence");
+    for (size_t point = 0; point < segmentation.cloud_->height * segmentation.cloud_->width;
+         point++)
+    {
+      unsigned int mx, my;
+      if (!worldToMap(*(iter_x + point), *(iter_y + point), mx, my))
+      {
+        RCLCPP_DEBUG(logger_, "Computing map coords failed");
+        continue;
+      }
+      unsigned int index = getIndex(mx, my);
+      uint8_t class_id = *(iter_class + point);
+      // Check if no cost was defined for the given class
+      if (!class_map_.count(segmentation.class_map_[class_id]))
+      {
+        RCLCPP_DEBUG(logger_, "Cost for class id %i was not defined, skipping", class_id);
+        continue;
+      }
+      // Get the class name from the buffered segmentation and then gets its cost from the class map
+      costmap_[index] = class_map_[segmentation.class_map_[class_id]];
+      touch(*(iter_x + point), *(iter_y + point), min_x, min_y, max_x, max_y);
+    }
+  }
+}
+
+void SemanticSegmentationLayer::onFootprintChanged()
+{
+  RCLCPP_DEBUG(rclcpp::get_logger("nav2_costmap_2d"),
+               "SemanticSegmentationLayer::onFootprintChanged(): num footprint points: %lu",
+               layered_costmap_->getFootprint().size());
+}
+
+void SemanticSegmentationLayer::updateCosts(nav2_costmap_2d::Costmap2D& master_grid, int min_i,
+                                            int min_j, int max_i, int max_j)
+{
+  if (!enabled_)
+  {
+    return;
+  }
+
+  if (!current_ && was_reset_)
+  {
+    was_reset_ = false;
+    current_ = true;
+  }
+  if (!costmap_)
+  {
+    return;
+  }
+  switch (combination_method_)
+  {
+    case 0:  // Overwrite
+      updateWithOverwrite(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    case 1:  // Maximum
+      updateWithMax(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    default:  // Nothing
+      break;
+  }
+}
+
+void SemanticSegmentationLayer::syncSegmPointcloudCb(
+  const std::shared_ptr<const vision_msgs::msg::SemanticSegmentation>& segmentation,
+  const std::shared_ptr<const sensor_msgs::msg::PointCloud2>& pointcloud)
+{
+  // check if data has the right dimensions
+  if (segmentation->width * segmentation->height != pointcloud->width * pointcloud->height)
```

Comment:
Can we not support them if they're not aligned? That would be important to note in documentation that this requires pre-aligned pointclouds.

Reply:
It would be nice to support this in the future, many RGBD sensors will give you different resolution choices for RGB and D information.

Reply:
I think different resolutions can be supported if they are aligned, the mask could be down/up scaled to match the pointcloud resolution (I can do it if you think is reasonable). For not aligned pointclouds things will get more complicated, extra information should be passed to the plugin since it must do the alignment before filling the costmap. I will leave that in the documentation, maybe an issue can be created when the PR is merged to keep it on track